### ISSUE TITLE
Autocomplete is wrong

### ISSUE THREAD

**Issue by Oatall** [rust-lang/rust-analyzer#18572]
**rust-analyzer version**: 
rust-analyzer 0.3.2196-standalone (327ab2958 2024-11-24) 
and nightly: rust-analyzer 0.4.2200-standalone (ab448661d 2024-11-28) 

**rustc version**: 
rustc 1.83.0 (90b35a623 2024-11-26)
1.82 also

**editor or extension**: 
VSCodium + rust-analyzer-no-server.vsix  (friend: VSCode + ? This bug also exists)

**relevant settings**: 

**repository link (if public, optional)**: 

**code snippet to reproduce**: 
```rust
use std::{fs::File, io::{BufReader, Read}};

fn main() {
    if let Ok(file) = File::open("bug.txt") {
        let mut reader = BufReader::new(file);
        let mut mod_info = String::new();
        reader.read_to_string();
    }
}
```
When I use helix, I encountered the bug https://github.com/rust-lang/rust-analyzer/issues/18547,
![Image](https://github.com/user-attachments/assets/f2a5b2e8-2f1a-4e5a-bd8e-ce009129f66e)
I switched to VSCodium, it's Ok. I think that bug may be Helix's bug. Not working properly with new version of RA.

But now I encountered this bug, I think this bug may be RA's bug.
![Image](https://github.com/user-attachments/assets/3829ac55-9089-4769-b77b-5b298cca6f7a)
When I type mod and then select `&mut mod_info` and confirm, it becomes `mod&mmod_info`
code snippet to reproduce is above.

**Comment by mfontanini** [comment]
This bug seems to have been introduced in release [2024-11-18](https://github.com/rust-lang/rust-analyzer/releases/tag/2024-11-18) and makes rust-analyzer almost unusable. You can see in this gif how I scroll through the available list of completion items and as I walk over them they become something else. If I press `<tab>` on a suggestion, it will autocomplete the thing it turned into, e.g. in this scenario if I had accepted the `swap` that shows up first, I would have instead got an `AsRef` autocompleted.

Could this be caused by https://github.com/rust-lang/rust-analyzer/pull/18503? It seems to be dealing with this behavior exactly.

![Image](https://github.com/user-attachments/assets/a8feb3eb-f681-458d-9c37-24dd7799d327)

**Comment by flodiebold** [comment]
@mfontanini your bug is #18547. This issue here describes a different problem in the end, which is why I left it open instead of closing as duplicate.

**Comment by Veykril** [comment]
Regarding this issue, I think this is caused by our hack for mapiing between the adjusted syntax tree and the original one getting too out of sync. This will likely have to wait on our salsa upgrade + its database forking feature (once thats implemented), that would trivially solve this.

**Comment by A4-Tacks** [comment]
This seems to be a duplicate of #15139
